### ISSUE TITLE
Coordinate Creation Spec

### ISSUE THREAD

**Issue by jmilloy** [creare-com/podpac#54]
## Summary
Create a document that serves as the coordinate creation spec. This issue is concurrent with coordinate class naming (#53). Ideally, this issue precedes coordinate documentation (#4) and testing (#3) -- it would be nice to update the tests to match the spec, and then update the code (and docstrings) to pass the tests. This issue also precedes the user guide (#45).

## Background
Coordinate creation is a challenging issue because it needs to be sufficiently general in order to cover enough use cases while remaining both concise as well as sufficiently simple/concrete/usable for a new user. We have overhauled coordinate creation a few times in the past, and it's time to really settle on a spec.

## Current Status
Coordinate creation involves two main components:
 - creation of the coordinate values of a single dimension (currently the `~Coord` classes)
 - assembling one or more of these 1d coordinates together with relational information (currently the `Coordinate` class)

### develop
The `Coordinate` class initialization accepts either keyword arguments or a stacked dictionary, which define the coordinates values for each dimension (key) with either `~Coord` objects or shortcuts that are mapped to `~Coord` objects. Keys with an underscore encode relational information (stacking). All of the coordinate creation options are available in one place.

```
# a few options
Coordinate({'lat_lon':(Coord([1, 2, 3, 4]), Coord([10, 20, 30, 40])), time:Coord('2018-01-01')})
Coordinate({'lat_lon':([1, 2, 3, 4], [10, 20, 30, 40]), time:'2018-01-01'})
Coordinate(lat_lon=([1, 2, 3, 4], [10, 20, 30, 40]), time='2018-01-01')
```

### PR #44 
The `Coordinate` class initialization requires a stacked dictionary with coordinate values for each dimension (key) defined by `~Coord` objects. Keys with an underscore encode relational information (stacking). The stacked dictionary argument matches the `Coordinate.stacked_dict` property.

A `coordinate` helper function accepts keyword arguments which define the coordinate values for each dimension (key) with shortcuts that are mapped to `~Coord` objects.

```
Coordinate({'lat_lon':(Coord([1, 2, 3, 4]), Coord([10, 20, 30, 40])), time:Coord('2018-01-01')})
podpac.coordinate('lat_lon'=([1, 2, 3, 4], [10, 20, 30, 40]), time='2018-01-01')
```

### @mlshapiro 
@mlshapiro has suggested replacing the keyword arguments with a list of values and second argument for dimension labels. This is similar to xarray DataArray creation: the main data and its structure are provided as a positional argument, with labeling meta-data passed as a second keyword argument. The values would accept either `~Coord` objects or shortcuts that are mapped to `~Coord` objects.

```
Coordinate([([1, 2, 3, 4], [10, 20, 30, 40]), '2018-01-01'], dims=[('lat', 'lon'), 'time'])
```

### Additional considerations
We discussed the pros and cons of putting all of the options into `Coordinate` class initialization vs. using factory functions, with general support for a spec with (potentially multiple) factory functions for common use-cases. The `Coordinate` initialization could remain strict, explicit, and general while factory functions would be easy-to-use, concise, and extensible.

**Comment by jmilloy** [comment]
A few thoughts:
 - I think the version in develop is appealing because all of the coordinate creation options are in one place. However I think the code is hard to follow and hard to test, and the documentation and usage overwhelming.
 - I think my PR #44 is appealing because the Coordinate init is simple and explicit and the coordinate factory is flexible and concise, so that both are easy to read, document, and test. The explicit dictionary argument required in init makes sense because it exactly matches the stacked_dict property.
 - I think Marc's proposal is appealing in the way it mirrors xarray.DataArray init and avoids keyword arguments. The values argument matches `Coordinate.stacked_dict.values()` and the labels argument matches `Coordinate.stacked_dict.keys()` (with the additional change of stacked keys being changed from underscore-separated strings to tuples, which is a separate concern).

**Comment by jmilloy** [comment]
An example of another helper function:

Currently you can create stacked uniform coordinates several ways.

```
Coordinate('lat_lon'=(coord_linspace(0, 1, 10), coord_linspace(0, 2, 10)))
Coordinate('lat_lon'=((0, 1, 10), (0, 2, 10)))
Coordinate('lat_lon'=((0, 1), (0, 2), 10))
```

The first two use a tuple matching the dimensions, with the second using the existing shortcut that maps to coord_linspace. The third one is an entirely separate case requiring a separate code path (plus tests and documentation) in coordinate creation. Consider replacing it with a helper function `stacked_linspace` that returns a tuple of UniformCoord objects (just like `coord_linspace` returns a single UniformCoord object).

```
Coordinate('lat_lon'=stacked_linspace((0, 1), (0, 2), 10))
```

I don't know, I was sort of just trying to think of another helper function. Maybe the names aren't quite right, or `coord_linspace` itself should handle both cases. But I thought it was worth writing down.

**Comment by mpu-creare** [comment]
I added a stub for the spec here:  4fa87abded5f4890ba6ed5227b5fb0339c41fc89
Jeff, have a go at it. Half time on this, half time on the unit tests. 
I'll try to get to your PR later today or tomorrow.

**Comment by jmilloy** [comment]
@mpu-creare @mlshapiro I pushed an updated (partial) spec to develop. I will probably flesh it out a little bit more on the airplane tomorrow, and possibly here and there I'm out the next two weeks, so take a peek again and let me know what you think.

Still need to do:
 - Coordinates methods
 - rotated coordinates
 - default ctype, etc handling (I get why the current solution doesn't really work anymore now)
 - coordinate evaluation rules, with regard to native coordinates (though this may more properly go in the node spec)

I'm particularly interested in the two downsampling examples at the bottom. The question applies for any indexing of stacked coordinates. It suggests to me that we should have a lightweight class `StackedCoordinates`. It would behave like a dictionary or tuple of Coordinate1D objects but also spit out the combined dims name (with underscores), the size, etc and handle mapping intersections, indexing, etc. We would probably still have the `stacked` function which would just alias StackedCoordinates, and the `stacked_linspace` helper function would return a StackedCoordinates as well. Is that something I should pursue? @mpu-creare do you anticipate any major cons, from your experience?

**Comment by jmilloy** [comment]
@mpu-creare Do you want to continue this issue as the coordinate refactor that implements the spec as well? Or should I start a new issue for that, and try to wrap the spec with the notes from our in-person meeting so we can close this?

**Comment by mpu-creare** [comment]
Start a new issue... in fact, you could do a WIP PR. 
Do please wrap in the notes from above so they don't get lost.
